src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -932,6 +934,228 @@ case class MatrixAggregateRowsByKey(child: MatrixIR, expr: IR) extends MatrixIR
   }
 }
 
+case class MatrixGroupColsByKey(child: MatrixIR, aggIR: IR) extends MatrixIR {
+  def children: IndexedSeq[BaseIR] = Array(child, aggIR)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): MatrixGroupColsByKey = {
+    newChildren match {
+      case Seq(child: MatrixIR, aggExpr: IR) =>
+        MatrixGroupColsByKey(child, aggExpr)
+    }
+  }
+
+  val typ = {
+    val newEntryType = aggIR.typ
+    child.typ.copyParts(entryType = coerce[TStruct](newEntryType), colType = child.typ.colKeyStruct)
+  }
+
+  def execute(hc: HailContext): MatrixValue = {
+    val mv = child.execute(hc)
+
+    // local things for serialization
+    val oldNCols = mv.nCols
+    val oldRVRowType = mv.typ.rvRowType
+    val oldColsType = TArray(mv.typ.colType)
+    val oldColValues = mv.colValues
+    val oldColValuesBc = mv.colValues.broadcast
+    val oldGlobalsBc = mv.globals.broadcast
+    val oldGlobalsType = mv.typ.globalType
+
+    val newRVType = typ.rvRowType
+    val newColType = typ.colType
+    val newEntriesIndex = typ.entriesIdx
+
+    val keyIndices = mv.typ.colKey.map(k => mv.typ.colType.field(k).index)
+    val keys = oldColValuesBc.value.map { a => Row.fromSeq(keyIndices.map(a.asInstanceOf[Row].get)) }.toSet.toArray
+    val nKeys = keys.length
+    val newColValues = oldColValues.copy(value = keys, t = TArray(newColType))
+
+    val columnMap = {
+      val keysByColumn = oldColValues.value.map { sa => Row.fromSeq(keyIndices.map(sa.asInstanceOf[Row].get)) }
+      val keyMap = keys.zipWithIndex.toMap
+
+      keysByColumn.map { k => if (k == null) -1 else keyMap(k) }
+        .zipWithIndex.map { case (newIndex, oldIndex) => (oldIndex, newIndex) }.toMap
+    }
+    val columnMapType = TDict(TInt32(), TInt32())
+
+    val transformInitOp: (Int, IR) => IR = { (nAggs, initOpIR) =>
+      val colIdx = ir.genUID()
+
+      def rewrite(x: IR): IR = {
+        x match {
+          case InitOp(i, args, aggSig) =>
+            InitOp(ir.ApplyBinaryPrimOp(ir.Add(),
+              ir.ApplyBinaryPrimOp(
+                ir.Multiply(),
+                ir.DictGet(ir.Ref("colMap", columnMapType), ir.Ref(colIdx, TInt32())),
+                ir.I32(nAggs)),
+              i),
+              args,
+              aggSig)
+          case _ =>
+            ir.Recur(rewrite)(x)
+        }
+      }
+
+      ir.ArrayFor(
+        ir.ArrayRange(ir.I32(0), ir.I32(oldNCols), ir.I32(1)),
+        colIdx,
+        rewrite(initOpIR))
+    }
+
+    val transformSeqOp: (Int, IR) => IR = { (nAggs, seqOpIR) =>
+      val colIdx = ir.genUID()
+
+      def rewrite(x: IR): IR = {
+        x match {
+          case SeqOp(a, i, aggSig) =>
+            SeqOp(a,
+              ir.ApplyBinaryPrimOp(ir.Add(),
+                ir.ApplyBinaryPrimOp(
+                  ir.Multiply(),
+                  ir.DictGet(ir.Ref("colMap", columnMapType), ir.Ref(colIdx, TInt32())),
```

Comment:
this is going to be super inefficient, we should have used the sample index instead